### PR TITLE
Speeding up importing of activities

### PR DESCRIPTION
**Pull request by rafacla**
Dears,

Adding a auxiliary list to avoid repetitive access to Yahoo API during importing process I was able to import 611 activities in my local testing machine, this may speed up the process specially for those that make recurring purchases.

### CODE REVIEW

File: apps/api/src/app/import/import.service.ts [hunk 1]
```diff
@@ -555,11 +555,16 @@ export class ImportService {
       }
     );
 
+    var listOfValidatedAssets: string[] = [];
```
Comment (dtslvr) [inline apps/api/src/app/import/import.service.ts:558]: I think a map is more intuitive for this kind of check:
```
const validatedAssetsMap: { [assetKey: string]: boolean } = {};
```

File: apps/api/src/app/import/import.service.ts [hunk 2]
```diff
@@ -580,12 +585,13 @@ export class ImportService {
...
         ) {
           throw new Error(
-            `activities.${index}.currency ("${currency}") does not match with "${assetProfile.currency}" and no exchange rate is available from "${currency}" to "${assetProfile.currency}"`
+            `activities.${index}.symbol("${symbol}").currency ("${currency}") does not match with "${assetProfile.currency}" and no exchange rate is available from "${currency}" to "${assetProfile.currency}."`
           );
         }
 
         assetProfiles[getAssetProfileIdentifier({ dataSource, symbol })] =
           assetProfile;
+        listOfValidatedAssets.push(symbol);
```
Comment (dtslvr) [inline apps/api/src/app/import/import.service.ts:594]: `validatedAssetsMap[getAssetProfileIdentifier({ dataSource, symbol })] = true;`

File: apps/api/src/app/import/import.service.ts [hunk 3]
```diff
@@ -555,11 +555,16 @@ export class ImportService {
...
+    var listOfValidatedAssets: string[] = [];
+
     for (const [
       index,
       { currency, dataSource, symbol }
     ] of uniqueActivitiesDto.entries()) {
-      if (dataSource !== 'MANUAL') {
+      if (
+        dataSource !== 'MANUAL' &&
+        listOfValidatedAssets.includes(symbol) === false
```
Comment (dtslvr) [inline apps/api/src/app/import/import.service.ts:566]: `!validatedAssetsMap[getAssetProfileIdentifier({ dataSource, symbol })]`

File: apps/api/src/app/import/import.service.ts [hunk 4]
```diff
@@ -580,12 +585,13 @@ export class ImportService {
           )
         ) {
           throw new Error(
-            `activities.${index}.currency ("${currency}") does not match with "${assetProfile.currency}" and no exchange rate is available from "${currency}" to "${assetProfile.currency}"`
+            `activities.${index}.symbol("${symbol}").currency ("${currency}") does not match with "${assetProfile.currency}" and no exchange rate is available from "${currency}" to "${assetProfile.currency}."`
```
Comment (dtslvr) [inline apps/api/src/app/import/import.service.ts:588]: This looks strange 😮

**Review by dtslvr — CHANGES_REQUESTED** [review]
Great optimization, @rafacla! 🙌🏻 

I have added some comments.

File: apps/api/src/app/import/import.service.ts (hunk 4 above)
Comment (rafacla) [reply apps/api/src/app/import/import.service.ts:588]: ooops, this one was an initial change I had made to identify which item was not being found at Yahoo, but I saw you already had fixed it a few lines before, we can drop this one.

**Review by dtslvr — CHANGES_REQUESTED** [review]
I'm confused. There was already an asset profiles map (see `assetProfiles`). Shouldn't we just use that in the check instead of introducing `validatedAssetsMap`?

**Comment by dtslvr** [comment]
What happened now? I think you were on the right track @rafacla! 🙃

This could already have an effect, but you don't need to populate the map again.

```
if (dataSource !== 'MANUAL') {
```
to
```
if (dataSource !== 'MANUAL' && !assetProfiles[getAssetProfileIdentifier({ dataSource, symbol })]) {
```

Can you give it a last try?

**Comment by rafacla** [comment]
Hi Thomas,

Actually this PR closed itself, because I had to rebase my git and open different branches to deal with the memory leak of TZ before cotinuing with this one, because the memory leak was causing me problems to debug the code after importing the first activities (otherwise I'd need to reset the database).

But, yesterday I was reading the code more troughly and acutally your code is already filtering unique assets in the process, so the changes I had initially proposed would not improve it afterall, but the thing is, I had installed Ghostfol.io in a Raspberry Pi 4, so in the Pi my former code indeed improved the time of importing, but I'll try to replicate the cases in the Pi in this weekend to understand what indeed happened to improve the speed.

As the Pi hardware is way weaker than my laptop, any improvement has more impact on it than on my dev machine.

Thanks!

**Comment by dtslvr** [comment]
Ah okay, I see. It's impossible to reopen this PR. So please create a new one if you have some insights 🙂

**Pull request closed without merging**